File: TienVeFE/src/app/services/user.service.ts
import { Injectable } from '@angular/core';
import { catchError, firstValueFrom, map, Observable, of } from 'rxjs';
import { API_ROUTES, APIResponse, BaseHTTPClient, CONSTANTS } from '../utils';
import { UserDTO } from '../dtos';
import { CountObject } from '../utils/types';

@Injectable({
    providedIn: 'root',
})
export class UserService {
    /** Constructor */
    constructor(private _http: BaseHTTPClient) {}

    /**
     * Get user profile
     */
    public getProfile(): Observable<APIResponse<UserDTO | undefined>> {
        return this._http.get<UserDTO | undefined>(API_ROUTES.USER.GET_PROFILE).pipe(
            map((httpRes) => {
                if (!httpRes.body?.data) return APIResponse.error(CONSTANTS.ERR_INTERNAL_SERVER_ERROR);
                return APIResponse.success(httpRes.body.message, httpRes.body.data);
            }),
            catchError((err) => {
                if (APIResponse.is(err)) {
                    return of(APIResponse.error(err.message, undefined, err.errors));
                }
                return of(APIResponse.error(CONSTANTS.ERR_INTERNAL_SERVER_ERROR));
            }),
        );
    }

    public async getByEmailPhone(emailPhone: string) {
        try {
            const route = API_ROUTES.USER.GET_BY_EMAIL_PHONE.replace(':email_phone', String(emailPhone));
            const httpRes = await firstValueFrom(this._http.get(route));
            if (!httpRes.body) return APIResponse.error(CONSTANTS.ERR_INTERNAL_SERVER_ERROR, null);
            if (!httpRes.body.data) return APIResponse.error(httpRes.body.message, null);

            const item = UserDTO.fromJson(httpRes.body.data);
            return APIResponse.success(httpRes.body.message, item);
        } catch (err) {
            if (APIResponse.is(err)) {
                return APIResponse.error(err.message, null, err.errors);
            }
            return APIResponse.error(CONSTANTS.ERR_INTERNAL_SERVER_ERROR, null);
        }
    }

    public async getTodayDebtCount() {
        try {
            const route = API_ROUTES.USER.GET_TODAY_DEBT_COUNT;
            const httpRes = await firstValueFrom(this._http.get<CountObject>(route));
            if (!httpRes.body?.data) return APIResponse.error(CONSTANTS.ERR_INTERNAL_SERVER_ERROR, null);

            return APIResponse.success(httpRes.body.message, httpRes.body?.data);
        } catch (err) {
            if (APIResponse.is(err)) {
                return APIResponse.error(err.message, null, err.errors);
            }
            return APIResponse.error(CONSTANTS.ERR_INTERNAL_SERVER_ERROR, null);
        }
    }
}

File: TienVeApp/lib/services/user_service.dart
import 'dart:io';

import 'package:tien_ve/dtos/message_dto.dart';
import 'package:tien_ve/dtos/user_dto.dart';
import 'package:tien_ve/utils/api_response.dart';
import 'package:tien_ve/utils/constants.dart';
import 'package:tien_ve/utils/http.dart';

class UserService {
  ///Login
  ///
  ///[emailPhone]: email or phone number
  ///
  ///[password]: MD5 hashed
  static Future<APIResponse<LoginUserDTO?>> login(
    String emailPhone,
    String password,
  ) async {
    try {
      final http = BaseHTTPClient();
      final reqBody = <String, dynamic>{
        "email_phone": emailPhone,
        "password": password,
      };
      final response = await http.post(APIRoutes.AUTH_LOGIN, reqBody);
      if (!response.isSuccess || response.data == null) {
        return APIResponse.error(message: response.message, data: null);
      }

      final data = LoginUserDTO.fromJson(response.data);
      return APIResponse.success(data: data);
    } catch (e) {
      return APIResponse.error();
    }
  }
}
